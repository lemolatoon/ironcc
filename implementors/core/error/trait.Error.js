(function() {var implementors = {
"ironcc":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.67.0/core/error/trait.Error.html\" title=\"trait core::error::Error\">Error</a> for <a class=\"struct\" href=\"ironcc/error/struct.CompileError.html\" title=\"struct ironcc::error::CompileError\">CompileError</a>"]]
};if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()