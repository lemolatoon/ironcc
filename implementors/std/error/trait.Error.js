(function() {var implementors = {};
implementors["ironcc"] = [{"text":"impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.64.0/std/error/trait.Error.html\" title=\"trait std::error::Error\">Error</a> for <a class=\"struct\" href=\"ironcc/error/struct.CompileError.html\" title=\"struct ironcc::error::CompileError\">CompileError</a>","synthetic":false,"types":["ironcc::error::CompileError"]}];
if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()