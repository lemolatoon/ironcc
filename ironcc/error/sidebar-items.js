window.SIDEBAR_ITEMS = {"enum":[["AnalyzeErrorKind",""],["CompileErrorKind",""],["GenerateErrorKind",""],["NotAllowedStmtKind",""],["ParseErrorKind",""],["TokenizeErrorKind",""],["Tokens",""],["TypeErrorKind",""],["TypeExpectedFailedKind",""],["UnexpectedTypeSizeStatus",""],["VariableKind",""]],"struct":[["CompileError",""]]};