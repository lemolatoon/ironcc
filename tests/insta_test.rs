use insta::assert_debug_snapshot;
pub mod test_utils;
use ironcc::error::CompileError;
use test_utils::CachedProcessor;

macro_rules! all {
    ($self: ident) => {
        assert_debug_snapshot!($self.tokens().as_ref().unwrap());
        assert_debug_snapshot!($self.program().as_ref().unwrap());
        assert_debug_snapshot!($self.conv_program().as_ref().unwrap());
    };
}

#[test]
fn insta_tests() -> Result<(), CompileError> {
    let src = "\nint main() {\nint i;\ni = 5;\nint* p; p = &i;\nint *p2; p2 = p + i;\n}";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn initializer() -> Result<(), CompileError> {
    let src = "int main() {int a = 5; int *p = &a; return 0;}";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn size_of() -> Result<(), CompileError> {
    let src = "\n\
    int main() {\n \
        int a = 5;\n \
        int b = sizeof (a);\n \
        int c = sizeof (int);\n \
        int d = sizeof (int*);\n \
    }\n\
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn array_syntax_sugar() -> Result<(), CompileError> {
    let src = "\n\
    int main() {\n \
        int array[5];
        return array[0];
    }\n\
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn scopes() {
    let src = "\n\
    int main() {\n \
        int a;
        {int *k;}
        {int *i; {int k;}}
        int c[2];
    }\n\
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn implicit_cast() {
    let src = "
    int main() {
        int a = 2;
        char b = a;
        b = 4;
    }
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn add_char() {
    let src = "
    int main() {
        int a = 2;
        char b = a;
        a + b;
        b + a;
    }
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn string_literal() {
    let src = r#"
    int printf(char *msg);

    int main() {
        char *msg = "Hello World";
        printf(msg);
    }
    "#;
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn string_literal2() {
    let src = r#"
    int printf(char *msg);

    int main() {
        char *msg;
        msg = "Hello World";
        printf(msg);
    }
    "#;
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}
