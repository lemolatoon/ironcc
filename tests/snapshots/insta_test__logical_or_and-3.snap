---
source: tests/insta_test.rs
expression: tester.conv_program().as_ref().unwrap()
---
ConvProgram {
    components: [
        Func(
            ConvFuncDef {
                ty: Func(
                    Base(
                        Int,
                    ),
                    [],
                ),
                name: "main",
                args: [],
                body: Block(
                    [
                        Expr(
                            ConvExpr {
                                kind: Assign(
                                    ConvExpr {
                                        kind: LVar(
                                            LVar {
                                                offset: 4,
                                                ty: Base(
                                                    Int,
                                                ),
                                            },
                                        ),
                                        ty: Base(
                                            Int,
                                        ),
                                        pos: Position {
                                            n_char: 8,
                                            n_line: 1,
                                        },
                                    },
                                    ConvExpr {
                                        kind: Num(
                                            1,
                                        ),
                                        ty: Base(
                                            Int,
                                        ),
                                        pos: Position {
                                            n_char: 16,
                                            n_line: 1,
                                        },
                                    },
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 8,
                                    n_line: 1,
                                },
                            },
                        ),
                        Expr(
                            ConvExpr {
                                kind: Assign(
                                    ConvExpr {
                                        kind: LVar(
                                            LVar {
                                                offset: 8,
                                                ty: Base(
                                                    Int,
                                                ),
                                            },
                                        ),
                                        ty: Base(
                                            Int,
                                        ),
                                        pos: Position {
                                            n_char: 8,
                                            n_line: 2,
                                        },
                                    },
                                    ConvExpr {
                                        kind: Num(
                                            0,
                                        ),
                                        ty: Base(
                                            Int,
                                        ),
                                        pos: Position {
                                            n_char: 16,
                                            n_line: 2,
                                        },
                                    },
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 8,
                                    n_line: 2,
                                },
                            },
                        ),
                        Expr(
                            ConvExpr {
                                kind: Assign(
                                    ConvExpr {
                                        kind: LVar(
                                            LVar {
                                                offset: 12,
                                                ty: Base(
                                                    Int,
                                                ),
                                            },
                                        ),
                                        ty: Base(
                                            Int,
                                        ),
                                        pos: Position {
                                            n_char: 8,
                                            n_line: 3,
                                        },
                                    },
                                    ConvExpr {
                                        kind: Conditional {
                                            cond: ConvExpr {
                                                kind: Conditional {
                                                    cond: ConvExpr {
                                                        kind: LVar(
                                                            LVar {
                                                                offset: 4,
                                                                ty: Base(
                                                                    Int,
                                                                ),
                                                            },
                                                        ),
                                                        ty: Base(
                                                            Int,
                                                        ),
                                                        pos: Position {
                                                            n_char: 16,
                                                            n_line: 3,
                                                        },
                                                    },
                                                    then: ConvExpr {
                                                        kind: LVar(
                                                            LVar {
                                                                offset: 8,
                                                                ty: Base(
                                                                    Int,
                                                                ),
                                                            },
                                                        ),
                                                        ty: Base(
                                                            Int,
                                                        ),
                                                        pos: Position {
                                                            n_char: 21,
                                                            n_line: 3,
                                                        },
                                                    },
                                                    els: ConvExpr {
                                                        kind: Num(
                                                            0,
                                                        ),
                                                        ty: Base(
                                                            Int,
                                                        ),
                                                        pos: Position {
                                                            n_char: 16,
                                                            n_line: 3,
                                                        },
                                                    },
                                                },
                                                ty: Base(
                                                    Int,
                                                ),
                                                pos: Position {
                                                    n_char: 16,
                                                    n_line: 3,
                                                },
                                            },
                                            then: ConvExpr {
                                                kind: Num(
                                                    2,
                                                ),
                                                ty: Base(
                                                    Int,
                                                ),
                                                pos: Position {
                                                    n_char: 26,
                                                    n_line: 3,
                                                },
                                            },
                                            els: ConvExpr {
                                                kind: Num(
                                                    0,
                                                ),
                                                ty: Base(
                                                    Int,
                                                ),
                                                pos: Position {
                                                    n_char: 16,
                                                    n_line: 3,
                                                },
                                            },
                                        },
                                        ty: Base(
                                            Int,
                                        ),
                                        pos: Position {
                                            n_char: 16,
                                            n_line: 3,
                                        },
                                    },
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 8,
                                    n_line: 3,
                                },
                            },
                        ),
                        Expr(
                            ConvExpr {
                                kind: Assign(
                                    ConvExpr {
                                        kind: LVar(
                                            LVar {
                                                offset: 16,
                                                ty: Base(
                                                    Int,
                                                ),
                                            },
                                        ),
                                        ty: Base(
                                            Int,
                                        ),
                                        pos: Position {
                                            n_char: 8,
                                            n_line: 4,
                                        },
                                    },
                                    ConvExpr {
                                        kind: Conditional {
                                            cond: ConvExpr {
                                                kind: Conditional {
                                                    cond: ConvExpr {
                                                        kind: LVar(
                                                            LVar {
                                                                offset: 4,
                                                                ty: Base(
                                                                    Int,
                                                                ),
                                                            },
                                                        ),
                                                        ty: Base(
                                                            Int,
                                                        ),
                                                        pos: Position {
                                                            n_char: 16,
                                                            n_line: 4,
                                                        },
                                                    },
                                                    then: ConvExpr {
                                                        kind: Num(
                                                            1,
                                                        ),
                                                        ty: Base(
                                                            Int,
                                                        ),
                                                        pos: Position {
                                                            n_char: 16,
                                                            n_line: 4,
                                                        },
                                                    },
                                                    els: ConvExpr {
                                                        kind: LVar(
                                                            LVar {
                                                                offset: 8,
                                                                ty: Base(
                                                                    Int,
                                                                ),
                                                            },
                                                        ),
                                                        ty: Base(
                                                            Int,
                                                        ),
                                                        pos: Position {
                                                            n_char: 21,
                                                            n_line: 4,
                                                        },
                                                    },
                                                },
                                                ty: Base(
                                                    Int,
                                                ),
                                                pos: Position {
                                                    n_char: 16,
                                                    n_line: 4,
                                                },
                                            },
                                            then: ConvExpr {
                                                kind: Num(
                                                    1,
                                                ),
                                                ty: Base(
                                                    Int,
                                                ),
                                                pos: Position {
                                                    n_char: 16,
                                                    n_line: 4,
                                                },
                                            },
                                            els: ConvExpr {
                                                kind: Num(
                                                    1,
                                                ),
                                                ty: Base(
                                                    Int,
                                                ),
                                                pos: Position {
                                                    n_char: 26,
                                                    n_line: 4,
                                                },
                                            },
                                        },
                                        ty: Base(
                                            Int,
                                        ),
                                        pos: Position {
                                            n_char: 16,
                                            n_line: 4,
                                        },
                                    },
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 8,
                                    n_line: 4,
                                },
                            },
                        ),
                    ],
                ),
                stack_size: 16,
            },
        ),
    ],
}
